### ISSUE TITLE
Best approach for dynamically changing batch sizes on iterable datasets

### ISSUE THREAD

**Issue by muellerzr** [pytorch/pytorch#128015]
### 🚀 The feature, motivation and pitch

The DeepSeek V2 paper proposed a training methodology where both the LR and the batch size were on a scheduler.

Exact description is below, however essentially:

* Increased total batch size from 2304 to 9216 over the first ~3% of training, or translated as a per-device bs of 288 -> 1152, starting pct of 25% -> 100%, over ~2.7% of the total steps before staying constantly at 9216
* They also used warmup-and-step-decay for the scheduler

![image](https://github.com/pytorch/pytorch/assets/7831895/76d41f50-cdcf-43c8-b9b9-23133ae52db4)

I'd like to have an approach that lets us dynamically modify the batch size on the fly (on a per-batch basis) for IterableDatasets. We can potentially do this via a wrapper `batch_sampler` (which accelerate can then wrap around to do per-device dispatching), however we'd need an approach that works for both iterable and non-iterable datasets to implement this fully. 

The FR is here in 🤗 Transformers: https://github.com/huggingface/transformers/issues/31222 

I know back in the day `fastai` let you just modify the `DataLoader` batch size on the fly, because of how they drew from their custom datasets. Would be really nice if we could have this in PyTorch in some capacity, to let open science flourish with all of the new ideas introduced in the DSv2 paper!

**Comment by andrewkho** [comment]
Thanks @muellerzr for the issue! I think @gokulavasan had an idea where this would be fairly simple to implement. Will follow up and get back to you.

**Comment by ifiaposto** [comment]
@muellerz  @gokulavasan @andrewkho @janeyx99 @ssnl @VitalyFedyunin @dzhulgakov 

here's a repo supporting this. let me know what you think :)

https://github.com/ifiaposto/hf_trainer_with_online_batch_size

ps making it compatible with accelerate would be so nice.

**Comment by ifiaposto** [comment]
> Thanks [@muellerzr](https://github.com/muellerzr) for the issue! I think [@gokulavasan](https://github.com/gokulavasan) had an idea that may work, although we need to think through different scenarios like when prefetch is enabled and whether that will give unpredictable results

prefetching seems tricky if someone wants the batch size to be synchronously updated. a provisional prefetching (which somehow would return the extra datapoints back to the iterator in case the batch size is set to a lower value than the current ) could be a treatment for that.  in my code, im currently manually disabling prefetching
